inst/include/Rcpp/String.h

```diff
@@ -329,11 +347,46 @@ namespace Rcpp {
             return std::wstring( s, s + strlen(s) );
         }
 
-
         inline const char* get_cstring() const {
             return buffer_ready ? buffer.c_str() : CHAR(data) ;
         }
 
+        inline const std::string get_encoding() const {
+            switch (enc) {
+                case CE_BYTES:
+                    return "bytes";
+                case CE_LATIN1:
+                    return "latin1";
+                case CE_UTF8:
+                    return "UTF-8";
+                default:
+                    return "unknown";
+            }
+        }
+
+        inline void set_encoding( cetype_t encoding ) {
+            enc = encoding;
+
+            if (valid) {
+                data = Rf_mkCharCE(Rf_translateCharUTF8(data), encoding);
```

Comment:
I don't think either of those speak to whether CHARSXPs are gc'd or not. This PR by @kevinushey suggests that they are: https://github.com/hadley/reshape/pull/40